core/crypto/crypto.c

```diff
@@ -0,0 +1,554 @@
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+TEE_Result crypto_mac_update(void *ctx __unused, uint32_t algo __unused,
+			     const uint8_t *data __unused, size_t len __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+
+TEE_Result crypto_mac_final(void *ctx __unused, uint32_t algo __unused,
+			    uint8_t *digest __unused,
+			    size_t digest_len __unused)
+{
+	return TEE_ERROR_NOT_IMPLEMENTED;
+}
+#endif /*_CFG_CRYPTO_WITH_MAC*/
+
+TEE_Result crypto_authenc_get_ctx_size(uint32_t algo __maybe_unused,
+				       size_t *size __maybe_unused)
+{
+	switch (algo) {
+#if defined(CFG_CRYPTO_CCM)
+	case TEE_ALG_AES_CCM:
+		*size = crypto_aes_ccm_get_ctx_size();
+		return TEE_SUCCESS;
+#endif
+#if defined(CFG_CRYPTO_GCM)
+	case TEE_ALG_AES_GCM:
+		*size = crypto_aes_gcm_get_ctx_size();
+		return TEE_SUCCESS;
+#endif
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+TEE_Result crypto_authenc_init(void *ctx __maybe_unused,
+			       uint32_t algo __maybe_unused,
+			       TEE_OperationMode mode __maybe_unused,
+			       const uint8_t *key __maybe_unused,
+			       size_t key_len __maybe_unused,
+			       const uint8_t *nonce __maybe_unused,
+			       size_t nonce_len __maybe_unused,
+			       size_t tag_len __maybe_unused,
+			       size_t aad_len __maybe_unused,
+			       size_t payload_len __maybe_unused)
+{
+	switch (algo) {
+#if defined(CFG_CRYPTO_CCM)
+	case TEE_ALG_AES_CCM:
+		return crypto_aes_ccm_init(ctx, mode, key, key_len, nonce,
+					   nonce_len, tag_len, aad_len,
+					   payload_len);
+#endif
+#if defined(CFG_CRYPTO_GCM)
+	case TEE_ALG_AES_GCM:
+		return crypto_aes_gcm_init(ctx, mode, key, key_len, nonce,
+					   nonce_len, tag_len);
+#endif
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+TEE_Result crypto_authenc_update_aad(void *ctx __maybe_unused,
+				     uint32_t algo __maybe_unused,
+				     TEE_OperationMode mode __unused,
+				     const uint8_t *data __maybe_unused,
+				     size_t len __maybe_unused)
+{
+	switch (algo) {
+#if defined(CFG_CRYPTO_CCM)
+	case TEE_ALG_AES_CCM:
+		return crypto_aes_ccm_update_aad(ctx, data, len);
+#endif
+#if defined(CFG_CRYPTO_GCM)
+	case TEE_ALG_AES_GCM:
+		return crypto_aes_gcm_update_aad(ctx, data, len);
+#endif
+	default:
+		return TEE_ERROR_NOT_IMPLEMENTED;
+	}
+}
+
+TEE_Result crypto_authenc_update_payload(void *ctx __maybe_unused,
+					 uint32_t algo __maybe_unused,
+					 TEE_OperationMode mode __maybe_unused,
+					 const uint8_t *src_data __maybe_unused,
+					 size_t src_len __maybe_unused,
+					 uint8_t *dst_data __maybe_unused,
+					 size_t *dst_len __maybe_unused)
+{
+	switch (algo) {
+#if defined(CFG_CRYPTO_CCM)
+	case TEE_ALG_AES_CCM:
+		if (*dst_len < src_len)
+			return TEE_ERROR_SHORT_BUFFER;
+		*dst_len = src_len;
+		return crypto_aes_ccm_update_payload(ctx, mode, src_data,
+						     src_len, dst_data);
+#endif
+#if defined(CFG_CRYPTO_GCM)
+	case TEE_ALG_AES_GCM:
+		if (*dst_len < src_len)
+			return TEE_ERROR_SHORT_BUFFER;
```

Comment:
should set *dst_len = src_len; before returning `TEE_ERROR_SHORT_BUFFER`.
here and below, and in few below functions.

Reply:
Thanks